Automattic/jetpack
# CRM: Remove $_SERVER['HTTP_USER_AGENT'] used by Zapier integration

This PR changes our API to remove `$_SERVER['HTTP_USER_AGENT']` usage. This affects only our Zapier integration, it currently relies on `$_SERVER['HTTP_USER_AGENT']` to determine if an API call is being done by Zapier or not.

We are removing `$_SERVER['HTTP_USER_AGENT']` because it is incompatible with WPVIP coding standards. With the changes from this PR we can achieve the same type of integration with Zapier, we are backwards compatible, and on top of that we add new functionality to our API.

## Proposed changes:

This PR introduces one new parameter for our customer_search API endpoint:
- `replace_hyphens_with_underscores_in_json_keys` - If set to `1` the json object keys returned by the CRM will have all hyphens replaced by underscores.

And also introduces one new parameter to all other endpoints:
- `external_api_name` - If set, this string will be used as the external api name in the CRM logs. (e.g. `EXTERNAL_API_NAME fired an API Action to update this contact`.

### Other information:


## Jetpack product discussion


https://github.com/Automattic/zero-bs-crm/issues/2772

## Does this pull request change what data or activity we track or use?



## Testing instructions:






You will have to call the API directly by using cUrl, RapidAPI, Insomnia, Postman, or similar tool.
(Alternatively  you can use Zapier if you know how to do it)

To call the API you need to pass the url (get) parameters `api_key` and `api_secret`. 
Generate them by enabling the API module (`wp-admin/admin.php?page=zerobscrm-modules`), then going to the API settings page (`wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=api`).
The API Docs are located here: https://automattic.github.io/jetpack-crm-api-docs/

- To test the API call `customer_search` 
    -  Create a custom field with at least two words (because we need a hyphen in the slug: for `test message` the slug is `test-message`).
    - Call `customer_search` without passing `replace_hyphens_with_underscores_in_json_keys`
        - You should get as response ``[{"id":"1","owner":"-1","status":"Lead", ... "test-message":"field value"}]``
    - Call `customer_search` passing `1` to `replace_hyphens_with_underscores_in_json_keys`
        - You should get as response ``[{"id":"1","owner":"-1","status":"Lead", ... "test_message":"field value"}]``

- To test the API call `create_company`, `create_customer`, `create_transaction`
    - Call the endpoint without passing  `external_api_name`
        - Check the log in the CRM, it should be a generic message informing that the object was created by an API call ( e.g. `Created from API Action`)
    - Call the endpoint passing some value to `external_api_name` ( e.g. `My Cool Service`)
        - Check the log in the CRM, it should be a message informing that the object was created by `My Cool Service` calling the API ( e.g. `Created by My Cool Service (API)`)
